compiler/z/codegen/SystemLinkagezOS.cpp

```diff
@@ -282,6 +282,91 @@ void TR::S390zOSSystemLinkage::createPrologue(TR::Instruction* cursor)
    setLastPrologueInstruction(cursor);
    }
 
+void
+TR::S390zOSSystemLinkage::setParameterLinkageRegisterIndex(TR::ResolvedMethodSymbol * method)
+   {
+   self()->setParameterLinkageRegisterIndex(method, method->getParameterList());
+   }
+
+void
+TR::S390zOSSystemLinkage::setParameterLinkageRegisterIndex(TR::ResolvedMethodSymbol * method, List<TR::ParameterSymbol> &parmList)
+   {
+   int32_t numGPRArgs = 0;
+   int32_t numFPRArgs = 0;
+   int32_t numVRFArgs = 0;
+
+   int32_t maxGPRArgs = self()->getNumIntegerArgumentRegisters();
+   int32_t maxFPRArgs = self()->getNumFloatArgumentRegisters();
+   int32_t maxVRFArgs = self()->getNumVectorArgumentRegisters();
+
+   ListIterator<TR::ParameterSymbol> paramIterator(&parmList);
+   for (TR::ParameterSymbol* paramCursor = paramIterator.getFirst(); paramCursor != NULL; paramCursor = paramIterator.getNext())
+      {
+      int32_t lri = -1;
+
+      switch (paramCursor->getDataType())
+         {
+         case TR::Int8:
+         case TR::Int16:
+         case TR::Int32:
+         case TR::Int64:
+         case TR::Address:
+            {
+            if (numGPRArgs < maxGPRArgs)
+               {
+               lri = numGPRArgs;
+               }
+
+            numGPRArgs++;
+            break;
+            }
+
+         case TR::Float:
+         case TR::Double:
+            {
+            if (numFPRArgs < self()->getNumFloatArgumentRegisters())
+               {
+               lri = numFPRArgs;
+               }
+            
+            // On 64-bit XPLINK floating point arguments leave "holes" in the GPR linkage registers, but not vice versa
+            numGPRArgs++;
+            numFPRArgs++;
+            break;
+            }
+
+         case TR::Aggregate:
+            {
+            TR_ASSERT_FATAL(false, "Support for aggregates is currently not implemented");
+            break;
+            }
+
+         case TR::VectorInt8:
+         case TR::VectorInt16:
+         case TR::VectorInt32:
+         case TR::VectorInt64:
+         case TR::VectorDouble:
+            {
+            if (numVRFArgs < self()->getNumVectorArgumentRegisters())
+               {
+               lri = numVRFArgs;
+               }
+
+            numVRFArgs++;
```

Comment:
Reading the Vendor Interfaces, I think the same change for FPRs above to bump numGPRArgs++ should be applicable for VRs as well.

Reply:
I couldn't find an updated reference for XPLINK with vector registers. Do they even participate in linkage conventions? How does one pass a vector register argument? If you have any links that would be appreciated. I'll make the change though.

Reply:
I was referencing: https://www-01.ibm.com/servers/resourcelink/svc00100.nsf/pages/zOSV2R3SA380688/$file/ceev100_v2r3.pdf around page 143/144, and the examples around page 911.

Reply:
Fixed in a03a579

Reply:
Ever consider the alignment requirement of passing Vector arguments, i.e.  leaving more holes in GPR side? or, is there alignment requirement at all?  I remembered I did all these on AIX side before, need to read the code in details to see why it failed now.